web-api/src/persistence/postgres/utils/transactions.ts

```diff
@@ -0,0 +1,46 @@
+import { settlePromises } from '@web-api/utilities/settlePromises';
+import { ConnectionStore, getConnection } from '@web-api/getConnection';
+
+export async function withTransaction<T>(fn: () => Promise<T>): Promise<T> {
+  // If we are already in a transaction, continue like normal.
+  if (inTransaction()) return fn();
+
+  const db = await getConnection();
+
+  // Otherwise, we need to start a transaction. We let kysely take care of the begin/commit/rollback details
+  return db.transaction().execute(async trx => {
```

Comment:
If we are not a nested withTransaction, then go ahead and use the kysely API to checkout a connection and start a transaction. We'll then run the callback in a context that tells it what connection it is using, that it is in a transaction, etc.

Reply:
Note that kysely is handling the "locking" for us to avoid a single connection trying to spin up two transactions at once. Process 2 will wait until process 1 has completed its transaction.